herd/AArch64Sem.ml

```diff
@@ -2382,7 +2331,67 @@ module Make
         (* | I_BL _|I_BLR _|I_BR _|I_RET _ *)
         | I_LD1M _|I_ST1M _) as i ->
             Warn.fatal "illegal instruction: %s" (AArch64.dump_instruction i)
-        )
+
+(* At the moment, NOP is the onely instruction that can be used
+ * to overwrite code
+ *)
+
+      let check_self ii =
+        let inst = ii.A.inst in
+        if not (AArch64Base.is_overwritable inst) then
+          do_build_semantics inst ii
+        else match Label.norm ii.A.labels with
+        | None -> do_build_semantics inst ii
+        | Some hd ->
+           let inst_val = V.instructionToV inst in
+           (* Shadow default control sequencing operator *)
+           let(>>*=) = M.bind_control_set_data_input_first in
+           let a_v = make_label_value ii.A.fetch_proc hd in
+           let a = (* Normalised address of instruction *)
+             A.Location_global (make_label_value ii.A.fetch_proc hd) in
+             read_loc_instr a ii
+             >>= fun actual_val ->
+             M.op Op.Eq actual_val inst_val
+             >>==
+               fun cond ->
+                 M.choiceT cond
+                   (commit_pred ii
+                    >>*= fun () -> M.mk_singleton_es (Act.NoAction) ii
+                    >>= fun () ->
+                      do_build_semantics inst ii )
+                   begin
+                     let mfail =
+                       let (>>!) = M.(>>!) in
+                       let m_fault =
+                         mk_fault
+                           a_v Dir.R AArch64.N ii
+                           (Some FaultType.AArch64.IllegalInstruction)
+                           (Some "Invalid") in
+                       commit_pred ii
+                       >>*= fun () -> m_fault >>| set_elr_el1 ii
+                       >>! B.Fault Dir.R in
+                     match inst with
+                     | AArch64Base.I_NOP -> mfail
```

Comment:
Is this correct? I understand that we would get here is the instruction was initially a NOP and it has been overwritten. What if the new instruction is valid (for example, a branch)?

Reply:
It is in the particular case we are handling at the moment: NOP is the only instruction that can be used to overwrite code. Hence if the initial instruction is NOP and that the actual is something else, this is faillure.

I'll comment more on this in the toplevel conversation

Reply:
yes, thank you!